drivers/ipm/ipm_sedi.c

```diff
@@ -0,0 +1,298 @@
+			sedi_ipc_read_dbl(device, &drbl_in);
+			len = IPC_HEADER_GET_LENGTH(drbl_in);
+			sedi_ipc_read_msg(device, ipm->incoming_data_buf, len);
+			ipm->rx_msg_notify_cb(dev,
+					      ipm->rx_msg_notify_cb_data,
+					      drbl_in, ipm->incoming_data_buf);
+		} else {
+			LOG_WRN("no handler for ipm new msg");
+		}
+		break;
+	case SEDI_IPC_EVENT_MSG_PEER_ACKED:
+		if (atomic_test_bit(&ipm->status, IPM_WRITE_IN_PROC_BIT)) {
+			k_sem_give(&ipm->device_write_msg_sem);
+		} else {
+			LOG_WRN("no sending in progress, got an ack");
+		}
+		break;
+	default:
+		return;
+	}
+}
+
+static int ipm_init(const struct device *dev)
+{
+	/* allocate resource and context*/
+	const struct ipm_sedi_config_t *info = dev->config;
+	sedi_ipc_t device = info->ipc_device;
+	struct ipm_sedi_context *ipm = dev->data;
+
+	info->irq_config();
+	k_sem_init(&ipm->device_write_msg_sem, 0, 1);
+	k_mutex_init(&ipm->device_write_lock);
+	ipm->status = 0;
+
+	sedi_ipc_init(device, ipm_event_dispose, (void *)dev);
+	atomic_set_bit(&ipm->status, IPM_PEER_READY_BIT);
+	LOG_DBG("ipm driver initialized on device: %p", dev);
+	return 0;
+}
+
+static int ipm_send_isr(const struct device *dev,
+			 uint32_t drbl,
+			 const void *msg,
+			 int msg_size)
+{
+	const struct ipm_sedi_config_t *info = dev->config;
+	sedi_ipc_t device = info->ipc_device;
+	uint32_t drbl_acked = 0;
+
+	sedi_ipc_write_msg(device, (uint8_t *)msg,
+			   (uint32_t)msg_size);
+	sedi_ipc_write_dbl(device, drbl);
+	do {
+		sedi_ipc_read_ack_drbl(device, &drbl_acked);
+	} while ((drbl_acked & BIT(IPC_BUSY_BIT)) == 0);
+
+	return 0;
+}
+
+static int ipm_sedi_send(const struct device *dev,
+		     int wait,
+		     uint32_t drbl,
+		     const void *msg,
+		     int msg_size)
+{
+	__ASSERT((dev != NULL), "bad params\n");
+	const struct ipm_sedi_config_t *info = dev->config;
+	struct ipm_sedi_context *ipm = dev->data;
+	sedi_ipc_t device = info->ipc_device;
+	int ret = -1;
+
+	/* check params, check status */
+	if ((msg_size > IPC_DATA_LEN_MAX)
+	    || ((msg_size > 0) && (msg == NULL))
+	    || ((drbl & BIT(IPC_BUSY_BIT)) == 0)) {
+		LOG_ERR("bad params when sending ipm msg on device: %p",
+			dev);
+		return -EINVAL;
+	}
+
+	if (wait == 0) {
+		LOG_ERR("not support no wait mode when sending ipm msg");
+		return -ENOTSUP;
+	}
+
+	if (k_is_in_isr()) {
+		return ipm_send_isr(dev, drbl, msg, msg_size);
+	}
+
+	k_mutex_lock(&ipm->device_write_lock, K_FOREVER);
+	set_ipm_dev_busy(dev, true);
+
+	if (!atomic_test_bit(&ipm->status, IPM_PEER_READY_BIT)) {
+		LOG_WRN("peer is not ready");
+		goto write_err;
+	}
+
+	/* write data regs */
+	if (msg_size > 0) {
+		ret = sedi_ipc_write_msg(device, (uint8_t *)msg,
+					 (uint32_t)msg_size);
+		if (ret != SEDI_DRIVER_OK) {
+			LOG_ERR("ipm write data fail on device: %p", dev);
+			goto write_err;
+		}
+	}
+
+	atomic_set_bit(&ipm->status, IPM_WRITE_IN_PROC_BIT);
+	/* write drbl regs to interrupt peer*/
+	ret = sedi_ipc_write_dbl(device, drbl);
+
+	if (ret != SEDI_DRIVER_OK) {
```

Comment:
It looks like here you're setting `ret` to a value coming from a completely different domain (`SEDI_DRIVER_*`). I don't think that's a good idea since then the caller of this function can't effectively interpret the return value. Either return `0` for success and `-errno` (POSIX error) for failure, or then always return SEDI error values, but not a mix of these.

Reply:
Done